### PR TITLE
Add "AddParensToTypedExpr" quickfix

### PR DESCRIPTION
**Pull request by DedSec256**
JetBrains

- [x] Support type casts

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+open JetBrains.ReSharper.Resources.Shell
+
+type AddParensToTypeTestFix(error: RuntimeCoercionSourceSealedError) =
+    inherit FSharpQuickFixBase()
+
+    let typeTestExpr = error.Expr.As<ITypeTestExpr>()
+    let prefixApp = if isNotNull typeTestExpr then typeTestExpr.Expression.As<IPrefixAppExpr>() else null
+    let testedExpr = if isNotNull prefixApp then prefixApp.ArgumentExpression else null
+
+    override x.Text = "Add parens to type test"
```
Comment (DedSec256) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs:19]: To be honest, this text does not seem clear enough to me. Maybe you have better ideas?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Resources.Shell
+
+type AddParensToTypeTestFix(error: RuntimeCoercionSourceSealedError) =
+    inherit FSharpQuickFixBase()
+
+    let typeTestExpr = error.Expr.As<ITypeTestExpr>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs:15]: Any reason not to get the expected type in the daemon process?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Highlightings/FcsErrors.xml [hunk 3]
```diff
@@ -320,4 +320,15 @@
     <QuickFix>ReplaceWithAbbreviatedTypeFix</QuickFix>
   </Error>
 
+  <Error staticGroup="FSharpErrors" name="RuntimeCoercionSourceSealed">
+    <Parameter type="IFSharpExpression" name="expr"/>
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Highlightings/FcsErrors.xml:324]: According to the quick fix code, shouldn't it be `ITypeTestExpr` instead?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Highlightings
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Resources.Shell
+
+type AddParensToTypeTestFix(error: RuntimeCoercionSourceSealedError) =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs:12]: Is this the only error type this quick fix is applicable to?

**Review by auduchinok — CHANGES_REQUESTED** [review]
Please check and/or add tests for:

* more args (including unexpected): `foo a b c :? T`
* different from/to types

Please also check if it's possible to reuse it for type cast expressions.

**Comment by DedSec256** [comment]
JetBrains
JetBrains1

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypedLikeExprFix.fs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    new (error: RuntimeCoercionSourceSealedError) =
+        AddParensToTypedLikeExprFix(error.Expr)
+
+    new (error: TypeConstraintMismatchError) =
+        AddParensToTypedLikeExprFix(error.Expr.As<ITypedLikeExpr>())
+
+    override x.Text =
+        match typedLikeExpr with
+        | :? IUpcastExpr -> "Add parens to type upcast"
+        | :? IDowncastExpr -> "Add parens to type downcast"
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypedLikeExprFix.fs:27]: Does distinguishing the cast kind bring anything useful?

**Review by auduchinok — COMMENTED** [review]
> Please check and/or add tests for:
>
> * more args (including unexpected): foo a b c :? T

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs (hunk 1 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypeTestFix.fs]: Not yet, unfortunately, although I don't like it much either.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Highlightings/FcsErrors.xml (hunk 3 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Highlightings/FcsErrors.xml:335]: Any update here?

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Highlightings/FcsErrors.xml:335]: I think this error applies not only to ITypeTestExpr.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/FcsErrorsStageProcessBase.fs [hunk 6]
```diff
@@ -158,6 +160,11 @@ type FcsErrorsStageProcessBase(fsFile, daemonProcess) =
         | FieldNotMutable ->
             createHighlightingFromNode FieldOrValueNotMutableError range
 
+        | RuntimeCoercionSourceSealed ->
+            match nodeSelectionProvider.GetExpressionInRange<IFSharpExpression>(fsFile, range, false, null) with
+            | :? ITypedLikeExpr as expr -> RuntimeCoercionSourceSealedError(expr, error.Message) :> _
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/FcsErrorsStageProcessBase.fs:165]: Why this type is not used in the type arguments list above?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypedLikeExprFix.fs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    override x.IsAvailable _ = isValid expr
+
+    override x.ExecutePsiTransaction _ =
+        use writeCookie = WriteLockCookie.Create(expr.IsPhysical())
+        use disableFormatter = new DisableCodeFormatter()
+
+        let exprCopy = expr.Copy()
+        let typedLikeExprCopy = typedLikeExpr.Copy()
+        let prefixApp = ModificationUtil.ReplaceChild(typedLikeExpr, prefixApp.Copy())
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/QuickFixes/AddParensToTypedLikeExprFix.fs:38]: Why copies of all the expressions are needed?

**Review by auduchinok — APPROVED** [review]

**Pull request merged**